### ISSUE TITLE
dict incompatible with TypedDict in direct yield

### ISSUE THREAD

**Issue by thilp** [microsoft/pyright#4631]
Hello! Thank you for Pyright. I believe it incorrectly reports a dict as incompatible with a TypedDict when that dict is directly used in a `yield` statement (without an intermediate, annotated variable):
```py
from typing import Iterator, TypedDict

class MyDict(TypedDict):
    id: int

def f() -> Iterator[MyDict]:
    yield {"id": 0}
```
```
pyright 1.1.293
/tmp/test.py
  /tmp/test.py:7:11 - error: Expression of type "dict[str, int]" cannot be assigned to yield type "MyDict"
    "dict[str, int]" is incompatible with "MyDict" (reportGeneralTypeIssues)
1 error, 0 warnings, 0 informations 
```
When assigning the same dict to a type-annotated variable, _and then_ yielding it, Pyright is happy:
```py
from typing import Iterator, TypedDict

class MyDict(TypedDict):
    id: int

def f() -> Iterator[MyDict]:
    d: MyDict = {"id": 0}
    yield d
```
```
pyright 1.1.293
0 errors, 0 warnings, 0 informations
```

I was put on the right track by [this comment](https://github.com/microsoft/pyright/issues/4390#issuecomment-1368713688).

**Comment by erictraut** [comment]
Thanks for the bug report. Pyright should be using bidirectional type inference to infer the TypedDict's type in this case. It was properly handling the case where the return type was a `Generator`, but not the case where the return type was an `Iterator`. This will be fixed in the next release.

**Comment by thilp** [comment]
Amazing, thank you! :partying_face:

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.294, which I just published. It will also be included in a future release of pylance.
